### DISCUSSION TITLE
MIDI actions with Program Changes Issue

### DISCUSSION THREAD

**Discussion by jmalon** [hydrogen-music/hydrogen#1690] [Hydrogen on Windows]
Hi,
I  would like to set up several MIDI events associated with Program Change messages but I can't get it to work.  When I enter a program number in the first parameter it is not saved, only it is saved  with I enter it in the second parameter of the MIDI System Tab.  In any case, it doesn't matter the number entered. The action is executed with any program change and I can't set up more than one event associated to a program change.
With the events associated to Control Changes it works perfectly but unfortunately the 4 CC# that I have available in my MIDI controller are already dedicated to other tasks.
I'm using the last stable version 1.1.1 and my OS is Windows 10

Any idea?
Thanks

**Comment by theGreatWhiteShark** [comment]
Could you check the [latest development version](https://ci.appveyor.com/api/buildjobs/tt1hs7odv22g1ssq/artifacts/Hydrogen-1.2.0-win64.exe)? I did a lot of changes in the MIDI tab and IIRC there was also a problem with PC messages in general which I fixed along the way.

**Reply by jmalon** [reply]
Thanks a lot,
but I have bad news. Now I can add several PROGRAM_CHANGE evets and they are saved in conf file. But the problem keeps going. I have watched there are a new celd:  (1 event par. and 3 more after Action List Box). No one of them are editables in the PROGRAM_CHAGE option. So , for example,  I created two events one for Play and the other for Stop. If I send whatever program change number the song plays, and then whatever number I send the song stops. 
I think that the ideal would be to be able to edit the first celd to enter the program number of the MIDI message Program Change and the systems would respond to this number. 
What is your idea?

**Reply by theGreatWhiteShark** [reply]
> No one of them are editables in the PROGRAM_CHAGE option.

You are right. We print out the value of the PC event in the logs but do not use it internally. I'll take care of it.

**Reply by theGreatWhiteShark** [reply]
Okay, wait. There is a reason can not assign an action to a specific PC value.

A CC or NoteOn message consists of a (note)parameter and a value. The former we use for mapping to specific actions and the latter as input variable for the action itself. The PC event, however, does only provide a single value. While it triggering toggling actions, like PLAY or STOP, using specific PC values would make sense, other actions expecting a numerical value as input would not be usable anymore. I'm afraid supporting event parameters in the MIDI table is not possible for PC events.

>  I created two events one for Play and the other for Stop. If I send whatever program change number the song plays, and then whatever number I send the song stops.

Yes, it is registered to the PC event in general and the actions run one after another. You could use PLAY/STOP_TOGGLE to switch between play and stop using one and the same event.

**Comment by jmalon** [comment]
Hi,
I have to read your explanation more carefully but I don't understand very
well the difficulty in using the Data Byte 1 of the Program Change MIDI
message.

Voice Message           Status Byte      Data Byte1          Data Byte2
-------------           -----------   -----------------   -----------------
Note on                       9x      Key number          Note on velocity
Control Change                Bx      Controller number   Controller value
Program Change                Cx      Program number      None

Effectively, Program Change doesn't use the 2nd Data Byte, but in Note
On I think that value (velocity) is not of much interest.

to launch an action.

**Reply by theGreatWhiteShark** [reply]
The MIDI actions in Hydrogen expect an integer input. Some of them, like stop or pause, do not use them. But, still, in order to exchange them seamlessly all of them expect such an input.

For `Note On` and `CC` events this integer value is retrieved from *Data Byte2* and *Data Byte1* is available and used to map to different actions. For `MMC` and `PC` events, however, only one number is submitted and *Data Byte1*  will be consumed as an input of the action function. That's why we do map the whole event to only a single action.

> but in Note On I think that value (velocity) is not of much interest.

Sure. The first value - the key number - can be used to map to a certain action and the second - velocity - serves as the input for the action function. I admit it's not too convenient to select a pattern number using note velocity. But mapping actions like stop or play to certain keys might be quite handy.

**Reply by jmalon** [reply]
First of all, thank you very much for your patience answering my questions. Sorry to be a bit boring, driven by my interest and curiosity to know the inner workings of things, and also by my passion for computer programming. I have actually programmed some audio but only with JUCE. 

That said, I'm afraid I'm going to have to take a look at the sources to properly understand why Data Byte1 can't be used to map actions in the PC and MMC events as it is in the Note On and CC events. It seems as if the roles of Data Byte1 end Data Byte2 are reversed. 
If you can give me some indication where to start looking I would appreciate it but without any commitment. I have already taken up enough of your time.

Thank you very much for your interest and congratulations for this great software that I recently discovered. 

PS.  I had to try it :)

**Reply by theGreatWhiteShark** [reply]
> If you can give me some indication where to start looking I would appreciate it but without any commitment. I have already taken up enough of your time.

Sure.

Main entry point is here: https://github.com/hydrogen-music/hydrogen/blob/master/src/gui/src/main.cpp
The MIDI table you can interact with in the preferences is here: https://github.com/hydrogen-music/hydrogen/blob/master/src/gui/src/Widgets/MidiTable.cpp
But main MIDI handling is done within the core: https://github.com/hydrogen-music/hydrogen/blob/master/src/core/MidiMap.h and https://github.com/hydrogen-music/hydrogen/blob/master/src/core/MidiAction.h
Incoming MIDI is handled here: https://github.com/hydrogen-music/hydrogen/blob/master/src/core/IO/MidiInput.cpp

**Comment by jmalon** [comment]
Thanks theGreatWhiteShark,
 I'm very busy these days but when I have a little free time I'll check it.
Happy New Year
